File: server/internal/daemon/knot_runtime_config.go
package daemon

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/multica-ai/multica/server/pkg/agent"
)

// knotRuntimeConfig is the schema the daemon expects under a knot or knot-http agent's
// `runtime_config` JSONB column. Every field is optional; an absent block (or
// the agent's whole runtime_config being null/empty) leaves the daemon-wide
// MULTICA_KNOT_AGENT_ID default in charge, so existing agents are unaffected.
//
// Schema:
//
//	{
//	  "knot": {
//	    "agent_id":    "ec4633074fe4413c83218e1f36b8e24d", // from `knot-cli list-agents`
//	    "client_uuid": "remote"                            // or a UUIDv4, or omit
//	  }
//	}
//
// It is nested under a "knot" key rather than sitting at the top level so one
// agent's runtime_config can carry blocks for unrelated concerns without the
// keys colliding — the same reason openclaw namespaces its gateway settings.
//
// Other providers' runtime_config payloads pass through untouched; this decoder
// only reads keys that have meaning for the knot backends.
type knotRuntimeConfig struct {
	Knot struct {
		AgentID string `json:"agent_id"`
		// ClientUUID selects WHICH registered machine runs the agent's tools
		// (chat_extra.agent_client_uuid). "remote" means "let Knot pick the
		// agent's own machine"; a UUIDv4 targets a specific one; empty pins the
		// local host. Only meaningful for knot-http.
		ClientUUID string `json:"client_uuid"`
	} `json:"knot"`
}

func supportsKnotRuntimeConfig(provider string) bool {
	return provider == "knot" || provider == "knot-http"
}

// decodeKnotRuntimeConfig extracts the per-agent Knot agent id from an agent's
// runtime_config payload, or "" when none is configured.
//
// Fails soft, deliberately: a malformed payload or a bad id logs a warning and
// returns "" so the daemon-wide default still applies, rather than blocking
// dispatch. The alternative would let one bad save break every task the agent
// runs — the same rule decodeOpenclawRuntimeConfig follows.
//
// The id is validated here rather than only in the backend because a wrong id
// is not a loud failure downstream: knot-cli silently substitutes its own
// default agent for an unknown id, so a typo would quietly bill and behave as a
// different agent. Rejecting it here means the run falls back to the configured
// default instead of an arbitrary one.
func decodeKnotRuntimeConfig(raw json.RawMessage, logger *slog.Logger) string {
	if len(raw) == 0 {
		return ""
	}
	var cfg knotRuntimeConfig
	if err := json.Unmarshal(raw, &cfg); err != nil {
		if logger != nil {
			logger.Warn("knot runtime_config: parse failed; falling back to the daemon-wide agent id", "error", err)
		}
		return ""
	}
	agentID := strings.TrimSpace(cfg.Knot.AgentID)
	if agentID == "" {
		return ""
	}
	if !agent.LooksLikeKnotAgentID(agentID) {
		if logger != nil {
			// The id is not a secret (it comes from `knot-cli list-agents`), so
			// logging it is what makes the typo findable.
			logger.Warn("knot runtime_config: agent_id is not a 32-character hex id; falling back to the daemon-wide agent id",
				"agent_id", agentID)
		}
		return ""
	}
	return agentID
}

// decodeKnotClientUUID validates the tool target before dispatch. Invalid
// configuration must never silently select another execution location.
func decodeKnotClientUUID(raw json.RawMessage, logger *slog.Logger) (string, error) {
	if len(raw) == 0 {
		return "", nil
	}
	var cfg knotRuntimeConfig
	if err := json.Unmarshal(raw, &cfg); err != nil {
		return "", fmt.Errorf("knot runtime_config: invalid client target configuration")
	}
	value := strings.TrimSpace(cfg.Knot.ClientUUID)
	if value == "" {
		return "", nil
	}
	if strings.EqualFold(value, "remote") {
		return "remote", nil
	}
	if !agent.LooksLikeKnotClientUUID(value) {
		return "", fmt.Errorf("knot runtime_config: client_uuid must be remote or a UUID")
	}
	return value, nil
}

File: packages/core/agents/knot-runtime-config.ts
// Knot-specific `runtime_config` schema.
//
// Stored under `agent.runtime_config` as freeform JSONB; only meaningful for
// agents whose runtime provider is knot or knot-http. The daemon decodes the same schema
// in `server/internal/daemon/knot_runtime_config.go` — keep both sides in
// lockstep when changing field names.
//
// Why this exists at all: knot-http puts the Knot agent id in its request URL,
// so a run cannot start without one. The daemon-wide MULTICA_KNOT_AGENT_ID
// cannot be set per agent (the daemon strips the whole MULTICA_* namespace from
// custom_env), so runtime_config is the per-agent channel. Resolution order at
// dispatch is:
//
//   KNOT_AGENT_ID (custom_env)  >  runtime_config  >  MULTICA_KNOT_AGENT_ID
//
// Note an Agent Builder carrier agent can use none of this: CreateAgentBuilder
// hardcodes runtime_config and custom_env to '{}', so those runs always fall
// through to the daemon-wide default.

export interface KnotRuntimeConfig {
  agentId?: string;
  // Which registered machine runs the agent's tools (chat_extra.agent_client_uuid,
  // knot-http only). "remote" = the agent's own machine; a UUIDv4 = a specific
  // one; absent = pin the local host. See resolveKnotHTTPClientUUID in
  // server/pkg/agent/knot_http.go — keep both sides in lockstep.
  clientUuid?: string;
}

// The sentinel that tells the backend to omit agent_client_uuid so Knot runs
// the agent on its own registered machine. Mirrors knotHTTPClientUUIDRemote.
export const KNOT_CLIENT_UUID_REMOTE = "remote";

// Length of the hex ids `knot-cli list-agents` reports.
const KNOT_AGENT_ID_LENGTH = 32;

// Mirrors knotLooksLikeAgentID in server/pkg/agent/knot.go: 32 lowercase hex
// characters. Validating in the UI matters because a wrong id fails quietly —
// knot-cli substitutes its own default agent for an unknown one, so a typo would
// bill and behave as a different agent instead of erroring.
export function looksLikeKnotAgentId(value: string): boolean {
  return (
    value.length === KNOT_AGENT_ID_LENGTH && /^[0-9a-f]+$/.test(value)
  );
}

// Parse an arbitrary runtime_config payload into the typed schema. Unknown keys
// are dropped and malformed payloads collapse to an empty object, so an invalid
// config renders as "nothing selected" rather than blocking the form with a
// parse error.
//
// An id that is present but malformed is deliberately KEPT here, unlike in the
// daemon which discards it. The form needs to show what is currently stored so
// the user can see and correct it; hiding it would make a bad saved value look
// like an empty field that mysteriously fails at dispatch.
export function parseKnotRuntimeConfig(raw: unknown): KnotRuntimeConfig {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return {};
  const root = raw as Record<string, unknown>;
  const knot = root.knot;
  if (!knot || typeof knot !== "object" || Array.isArray(knot)) return {};
  const block = knot as Record<string, unknown>;
  const out: KnotRuntimeConfig = {};
  if (typeof block.agent_id === "string" && block.agent_id.trim() !== "") {
    out.agentId = block.agent_id.trim();
  }
  // Kept even when malformed, for the same reason as agent_id: the form must
  // show what is stored so a bad value can be seen and corrected, not hidden.
  if (typeof block.client_uuid === "string" && block.client_uuid.trim() !== "") {
    out.clientUuid = block.client_uuid.trim();
  }
  return out;
}

// Render the typed form state back into the wire shape the API accepts.
//
// An unset id produces `{}` rather than `{knot:{agent_id:""}}` so clearing the
// picker actually falls back to the daemon-wide default instead of pinning an
// empty id that the daemon would have to special-case. The same holds for
// client_uuid: only fields the user actually set survive, and an all-empty
// selection drops the knot block entirely.
export function serializeKnotRuntimeConfig(
  cfg: KnotRuntimeConfig,
): Record<string, unknown> {
  const agentId = cfg.agentId?.trim() ?? "";
  const clientUuid = cfg.clientUuid?.trim() ?? "";
  const knot: Record<string, unknown> = {};
  if (agentId !== "") knot.agent_id = agentId;
  if (clientUuid !== "") knot.client_uuid = clientUuid;
  if (Object.keys(knot).length === 0) return {};
  return { knot };
}

// Stable equality across two parsed configs, for the form's dirty detector.
// Absent and empty are the same thing (both mean "use the default"), so
// toggling a selection on and off again is not a change.
export function knotRuntimeConfigEquals(
  a: KnotRuntimeConfig,
  b: KnotRuntimeConfig,
): boolean {
  return (
    (a.agentId ?? "") === (b.agentId ?? "") &&
    (a.clientUuid ?? "") === (b.clientUuid ?? "")
  );
}

// Merge a fully-formed Knot selection into an agent's existing runtime_config,
// preserving any other provider's block.
//
// Needed because the settings tab PUTs runtime_config wholesale: writing only
// the knot block would silently drop an openclaw gateway pin (or anything added
// later) from the same agent. An empty selection removes the knot block rather
// than storing blank fields.
function mergeKnotBlock(
  existing: unknown,
  next: KnotRuntimeConfig,
): Record<string, unknown> {
  const base: Record<string, unknown> =
    existing && typeof existing === "object" && !Array.isArray(existing)
      ? { ...(existing as Record<string, unknown>) }
      : {};
  const wire = serializeKnotRuntimeConfig(next);
  if (wire.knot) {
    base.knot = wire.knot;
  } else {
    delete base.knot;
  }
  return base;
}

// Set the Knot agent id while PRESERVING any client_uuid already stored on the
// same agent. Replacing the whole knot block (the old behavior) would silently
// wipe a configured client_uuid every time the agent picker changed.
export function withKnotAgentId(
  existing: unknown,
  agentId: string,
): Record<string, unknown> {
  const prev = parseKnotRuntimeConfig(existing);
  return mergeKnotBlock(existing, { ...prev, agentId: agentId.trim() });
}

// Set the Knot client_uuid (which machine runs the tools) while preserving any
// agent_id already stored. Pass "" to clear it (falls back to pinning the local
// host); "remote" to dispatch to the agent's own registered machine.
export function withKnotClientUuid(
  existing: unknown,
  clientUuid: string,
): Record<string, unknown> {
  const prev = parseKnotRuntimeConfig(existing);
  return mergeKnotBlock(existing, { ...prev, clientUuid: clientUuid.trim() });
}

// This is the saved requested target; custom_env may override it at dispatch.
// It never establishes the actual execution location of Knot file tools.
export function knotToolTarget(raw: unknown): "local" | "remote" | "client" | "invalid" {
  if (raw != null && (typeof raw !== "object" || Array.isArray(raw))) return "invalid";
  if (raw && typeof raw === "object" && "knot" in raw && raw.knot != null) {
    const knot = raw.knot;
    if (typeof knot !== "object" || Array.isArray(knot)) return "invalid";
    if ("client_uuid" in knot && knot.client_uuid != null && typeof knot.client_uuid !== "string") return "invalid";
  }
  const value = parseKnotRuntimeConfig(raw).clientUuid;
  if (!value) return "local";
  if (value.toLowerCase() === "remote") return "remote";
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(value) ? "client" : "invalid";
}
